src/main/java/vazkii/patchouli/client/book/text/TextLayouter.java

```diff
@@ -0,0 +1,199 @@
+package vazkii.patchouli.client.book.text;
+
+import net.minecraft.client.gui.FontRenderer;
+import vazkii.patchouli.client.book.gui.GuiBook;
+
+import java.text.BreakIterator;
+import java.util.ArrayList;
+import java.util.List;
+
+public class TextLayouter {
+	private final List<Word> words = new ArrayList<>();
+	private final GuiBook gui;
+	private final int pageX;
+	private final int lineHeight;
+	private final int pageWidth;
+
+	public TextLayouter(GuiBook gui, int pageX, int pageY, int lineHeight, int pageWidth) {
+		this.gui = gui;
+		this.pageX = pageX;
+		this.lineHeight = lineHeight;
+		this.pageWidth = pageWidth;
+
+		y = pageY;
+	}
+
+	private int y;
+	private List<Word> linkCluster = null;
+	private List<Span> spanCluster = null;
+
+	private List<SpanTail> pending = new ArrayList<>();
+	private int lineStart = 0;
+	private int widthSoFar = 0;
+
+	public void layout(List<Span> spans) {
+		StringBuilder allText = new StringBuilder();
+		for (Span span : spans)
+			allText.append(span.text);
+
+		BreakIterator breakIterator = BreakIterator.getLineInstance();
+		breakIterator.setText(allText.toString());
+
+		List<Span> paragraph = new ArrayList<>();
+		for (Span span : spans) {
+			if (span.lineBreaks > 0) {
+				layoutParagraph(paragraph);
+
+				widthSoFar = 0;
+				y += span.lineBreaks * lineHeight;
+				paragraph.clear();
+			}
+
+			paragraph.add(span);
+		}
+		if (!paragraph.isEmpty())
+			layoutParagraph(paragraph);
+	}
+
+	// a paragraph is a series of spans without explicit line break
+	private void layoutParagraph(List<Span> paragraph) {
+		String text = toString(paragraph);
+		BreakIterator iterator = BreakIterator.getLineInstance();
```

Comment:
Just curious - there is a [locale-sensitive version of `getLineInstance`](https://docs.oracle.com/javase/8/docs/api/java/text/BreakIterator.html#getLineInstance-java.util.Locale-), and `MinecraftForgeClient.getLocale` is available to retrieve current game language setting as a `java.util.Locale`, so is there any possibility to use `getLineInstance(Locale)` instead? I haven't dug into details behind this, but I figure that doing so won't hurt much.

Reply:
Oh, I thought Minecraft would set the current locale to the currently configured language; but yes, this can be changed.

Reply:
IIRC Minecraft sets current locale to system default (i.e. with respect to your OS locale, not the game setting) if it ever sets it. `MinecraftForgeClient.getLocale` was introduced in MinecraftForge/MinecraftForge#3522 as a way to get a `java.util.Locale` instance with respect to game settings for formatting (reads `java.util.Formatter`) purpose; the behavior got fixed twice in https://github.com/MinecraftForge/MinecraftForge/commit/f9b8c992eec06c0e8ea53d6f1989c117980642c3 and MinecraftForge/MinecraftForge#5105 respectively. 
btw, `TextLayouter::layout` also has a call to `BreakIterator::getLineInstance`. Mind you change that as well?